### ISSUE TITLE
Prioritizing IDs over Names

### ISSUE THREAD

**Issue by aluzzardi** [docker/docker#13286]
Docker should prioritize IDs over Names in order to mitigate CVE-2014-5278:
https://groups.google.com/forum/m/#!topic/docker-user/jyf9_mYcMI8

In #8087, prioritizing Names over IDs was proposed and rejected (rightfully so).

However, in #9311 that behavior was proposed again and merged.

We must:
- Revert that behavior
- Add a test with a comment linking to the CVE in order to make sure it doesn't happen again

This is the offender: https://github.com/docker/docker/blob/master/daemon/daemon.go#L128
```go
	// GetByName will match only an exact name provided; we ignore errors
	containerByName, _ := daemon.GetByName(prefixOrName)
	containerId, indexError := daemon.idIndex.Get(prefixOrName)

	if containerByName != nil {
		// prefix is an exact match to a full container Name
		return containerByName, nil
	}
```

ping @docker/core-maintainers @acbodine @ewindisch @thaJeztah @diogomonica

**Comment by tiborvass** [comment]
@aluzzardi why not error out when it's ambiguous?

**Comment by aluzzardi** [comment]
I believe IDs (as in, full exact IDs) are used mostly by machines. If someone names a container after my ID, the machine won't be able to handle the conflict resolution (as in, if you "shadow" my ID then the only way to interact with the container is by name).

This could possibly break swarm, compose, and pretty much anything that uses the API and is not a human being :)

**Comment by thaJeztah** [comment]
Thanks, @aluzzardi. For reference, this came up in https://github.com/docker/swarm/pull/803.

I don't think an ambiguous match (i.e. both an ID and name match) should error out, because that would still be a vulnerability; creating a container with a name that matches an ID, would block access to that container.

IMO the flow should basically be;

```
if exactMatchOnID {
    return match
}

// Fuzzy matching - do whatever is nice / useful / easy to use
```

**Comment by thaJeztah** [comment]
Added this to the 1.7 milestone so that the behavior of Swarm and Docker match and this is a regression

**Comment by ewindisch** [comment]
+1;  Exact matches to the ID should supersede all.

Note that #8069 preferred fuzzy id matching over names. As a result, we had complaints that a container named 'db' would conflict with a container id named 'db00000...' and could not be referred to by name. A test should also exist for this and considerations made, otherwise someone might follow up with another wrong fix.

**Comment by estesp** [comment]
The first comment in this issue generates confusion as it misses a key difference between #8087 and #9705 (which is the PR that closes #9311; 9311 is only an issue).

8087 was closed because it "broke" the CVE fix; it disregarded IDs over names.  9705 was a rework based on @ewindisch feedback that full IDs had to be preferred over names **always**, so that is the way the code operates today.  A test was added for this behavior which will break if it is ever changed.

The only change introduced is the allowance for **short IDs** of any length < 64 to be superseded by names because of the hiding issue.. a container named "db" will be unreachable if and when any docker daemon has a container (or image for certain commands) which starts with the prefix "db".

If the test is not sufficient to maintain this behavior we can review, but I remember reviewing it during the 9705 PR review and it was testing all cases at the time.

**Comment by thaJeztah** [comment]
@estesp is completely right; as can be seen at these lines; https://github.com/docker/docker/blob/b2a43baf2e2cc68c83383a7524441f81bc4c4725/daemon/daemon.go#L125-L128

```
func (daemon *Daemon) Get(prefixOrName string) (*Container, error) {
	if containerByID := daemon.containers.Get(prefixOrName); containerByID != nil {
		// prefix is an exact match to a full container ID
		return containerByID, nil
	}
```

So looks like this is a non-issue after all.. (And NO idea why I didn't see this, I guess I didn't check the link to the actual code :blush:

I'm going to close this issue, because there isn't an issue here (as far as I can see).
